packages/core/contracts/oracle/implementation/cross-chain/BeaconOracle.sol

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "../../interfaces/OracleAncillaryInterface.sol";
+import "../../../common/implementation/MultiRole.sol";
+import "../../interfaces/FinderInterface.sol";
+import "./BridgeInterface.sol";
+import "../Constants.sol";
+
+/**
+ * @title Simple implementation of the OracleInterface used to communicate price request data cross-chain between
+ * EVM networks. Can be extended either into a "Source" or "Sink" oracle. The intention is that the "Source" Oracle
+ * is the originator of price resolution data, secured by the DVM, and the "Sink" Oracle can both receive this price
+ * resolution data and request prices cross-chain to the "Source" Oracle. The "Sink" is designed to be deployed on
+ * non-Mainnet networks and the "Source" should be deployed on Mainnet.
+ */
+abstract contract BeaconOracle is OracleAncillaryInterface, MultiRole {
+    enum Roles {
+        Owner, // The owner manages the other roles.
+        Requester, // Can make price requests.
+        Publisher // Can publish price request results.
+    }
+
+    struct Price {
+        bool isAvailable;
+        int256 price;
+    }
+
+    // The two structs below are used in an array and mapping to keep track of prices that have been requested but are
+    // not yet available.
+    struct QueryIndex {
+        bool isValid;
+        uint256 index;
+    }
+
+    // Represents a (identifier, time) point that has been queried.
+    struct QueryPoint {
+        bytes32 identifier;
+        uint256 time;
+        bytes ancillaryData;
+    }
+
+    // Conceptually we want a (time, identifier, ancillaryData) -> price map.
+    mapping(bytes32 => mapping(uint256 => mapping(bytes => Price))) internal verifiedPrices;
+
+    // The mapping and array allow retrieving all the elements in a mapping and finding/deleting elements.
+    // Can we generalize this data structure?
+    mapping(bytes32 => mapping(uint256 => mapping(bytes => QueryIndex))) internal queryIndices;
+
+    QueryPoint[] public requestedPrices;
+
+    // Finder to provide addresses for DVM contracts.
+    FinderInterface public finder;
+
+    event PriceRequestAdded(address indexed requester, bytes32 indexed identifier, uint256 time, bytes ancillaryData);
+    event PushedPrice(
+        address indexed pusher,
+        bytes32 indexed identifier,
+        uint256 time,
+        bytes ancillaryData,
+        int256 price
+    );
+
+    /**
+     * @notice Constructor.
+     * @param _finderAddress finder to use to get addresses of DVM contracts.
+     */
+    constructor(address _finderAddress) public {
```

Comment:
Do we need to initialize the multirole roles in the constructor?

Reply:
Ah yes, good point.